src/main/java/org/jboss/threads/CachedExecutorView.java

```diff
@@ -0,0 +1,298 @@
+        // to have pending runnables.
+        return Collections.emptyList();
+    }
+
+    @Override
+    public boolean isShutdown() {
+        return isShutdown(stateUpdater.get(this));
+    }
+
+    private static boolean isShutdown(int state) {
+        return (state & SHUTDOWN_MASK) != 0;
+    }
+
+    @Override
+    public boolean isTerminated() {
+        return isTerminated(stateUpdater.get(this));
+    }
+
+    private static boolean isTerminated(int state) {
+        return (state & TERMINATED_MASK) != 0;
+    }
+
+    @Override
+    public boolean awaitTermination(long timeout, TimeUnit unit) throws InterruptedException {
+        long remainingNanos = unit.toNanos(timeout);
+        // Use the system precise clock to avoid issues resulting from time changes.
+        long now = System.nanoTime();
+        synchronized (shutdownLock) {
+            while (!isTerminated()) {
+                remainingNanos -= Math.max(-now + (now = System.nanoTime()), 0L);
+                long remainingMillis = TimeUnit.MILLISECONDS.convert(remainingNanos, TimeUnit.NANOSECONDS);
+                if (remainingMillis <= 0) {
+                    return false;
+                }
+                shutdownLock.wait(remainingMillis);
+            }
+            return true;
+        }
+    }
+
+    @Override
+    public void execute(Runnable task) {
+        Assert.checkNotNullParam("task", task);
+        boolean submittedTask = false;
+        int snapshot = stateUpdater.updateAndGet(this, incrementActiveStateUpdater);
+        try {
+            if (isShutdown(snapshot)) {
+                throw new RejectedExecutionException("Executor has been shut down");
+            }
+            delegate.execute(new CachedExecutorViewRunnable(task));
+            submittedTask = true;
+        } finally {
+            if (!submittedTask) {
+                decrementActive();
+            }
+        }
+    }
+
+    private void decrementActive() {
+        int stateSnapshot = stateUpdater.updateAndGet(this, decrementActiveStateUpdater);
+        if (isTerminated(stateSnapshot)) {
+            synchronized (shutdownLock) {
+                shutdownLock.notifyAll();
+            }
+        }
+    }
+
+    private static int getActiveCount(int state) {
+        return state & ACTIVE_COUNT_MASK;
+    }
+
+    private final class CachedExecutorViewRunnable implements Runnable {
+
+        private final Runnable delegate;
+
+        @Nullable
+        private volatile Thread thread;
+
+        CachedExecutorViewRunnable(Runnable delegate) {
+            this.delegate = delegate;
+        }
+
+        @Override
+        public void run() {
+            Thread currentThread = Thread.currentThread();
+            Set<CachedExecutorViewRunnable> runnables = activeRunnables;
+            this.thread = currentThread;
+
+            String originalName = currentThread.getName();
+            try {
+                currentThread.setName(nameFunction.rename(originalName));
```

Comment:
Rename doesn't need to be called each invocation, a threadlocal anchored to the CachedExecutorView instance should allow us to calculate these once per thread.

Reply:
Implementations of the name function could implement the threadlocal piece themselves as well, that way a an identity thread name function doesn't incur the cost of a map lookup.